### ISSUE TITLE
Registering a device

### ISSUE THREAD

**Issue by Ericmas001** [codetheweb/tuyapi#22]
Some discussion on #5 are about figuring out the registration process, I'm starting a thread here so we can focus on this particular issue.

I recorded with fiddler on my old phone the action of adding a new device to the SmartLife App. I also recorded with WireShark all the UDP packets sent by either the device I was registering or the phone I was using to register the device.

I will try to make something readable of all of this and post all the info here, i will just need a little time.

My conclusion from a quick glance a the data:

- During the process, the phone is making a constant repeated request to the SmartLifeAPI, while sending on the network in a loop a sequence of 4 UDP packets. (143 times)
- At some point, it changes he UDP loop. The phone starts sending a bunch of UDP packets ( a sequence of 64 different packets) in a loop. (30 times), while some other requests are made to the SmartLifeAPI

So my guess is: 
- The lost devices (with no connection to anything) are constantly listening and interested about protocol ABCDXYZ
- The first 4 UDP packets are something like "Hey, somebody listening, I got something to say about protocol ABCDXYZ. So the device will get interested to what the phone has to say
- The second UDP sequence of 64 packets are hte in fo, and I guess it contains: Wifi Information + SmartLifeApi information
- When the device got all the info, he then contact the SmartLifeAPI and register itself.
- During this time, the phone is constantly poking the SmartLifeAPI to know if the device has been registered yet. Then when it does, the app goes to 100% and asks you some info.

That's how i imagine the registration works.

**Comment by codetheweb** [comment]
From my own packet sniffing I can confirm this is correct.

Here's an idea: what if we're able to send a custom server address in the configuration step?

**Comment by Ericmas001** [comment]
That was my original Idea.
I had plans to try: Different WIFI ssid, different WIFI password, different pairing app (There is at leat 3 or 4 that we can try. With those tests I think we would be available to determine what is what.

**Comment by Ericmas001** [comment]
During registration, device makes
 - [One token.create request](https://github.com/Ericmas001/Tuya-Api-Tools/wiki/Requests%3A-adding-device#01-smgwtokencreate)
 - [Multiple unsuccessful list.token requests](https://github.com/Ericmas001/Tuya-Api-Tools/wiki/Requests%3A-adding-device#02-tuyamdevicelisttoken)
 - [Multiple successful list.token requests](https://github.com/Ericmas001/Tuya-Api-Tools/wiki/Requests%3A-adding-device#07-tuyamdevicelisttoken)

**Comment by Ericmas001** [comment]
I've registered 3 times the same device, to the same network, and I compared the UDP packet. I did not make any conclusion yet, only some highlighting. Maybe someone can make some sense with this 😃 
http://htmlpreview.github.io/?raw.githubusercontent.com/Ericmas001/Tuya-Api-Tools/master/SmartLife.htm

- Yellow columns are directly linked to the packet length. (some are not highlighted, because they are not the same each time)
- Other Colored columns are lightly colored when it's different for every tries, but all the same during the batch. If it's heavily colored, it changes during the batch too.
- Grey columns are constant between tries & batches

**Comment by Ericmas001** [comment]
Ok I've played a little with the APK.
The apps all uses with JNI a C++ `.so` library to Create and Send UDP packets. [jniLibs.zip](https://github.com/codetheweb/tuyapi/files/1684297/jniLibs.zip)
2 methods are exposed to JNI
```
package com.tuya.smart.android.device;

public class TuyaSmartLink
{
  static
  {
    try
    {
      System.loadLibrary("TuyaSmartLink");
      return;
    }
    catch (Throwable localThrowable)
    {
      localThrowable.printStackTrace();
    }
  }
  
  public static native void sendStatusStop();
  
  public static native int smartLink(String paramString1, String paramString2, String paramString3, int paramInt1, int paramInt2, int paramInt3, int paramInt4, int paramInt5);
}
```
Just remember that some info are obtained by the App at the beginning of the device registration process, the app says "I want to register a device", and the server sends back:
```
"secret": "0NiE",
"token": "QyMLEFzw"
```
The App also asks for the WifiPassword of the Wifi network that the device is connected to. This password, and the name of the wifi network, will be also used in the process.

Finally, the App determine the current region. There seems to be 3 regions, EU = Europe, AZ = America and AY = Asia. The complete code about choosing the region is there: [TyCommonUtil.zip](https://github.com/codetheweb/tuyapi/files/1684327/TyCommonUtil.zip)

So for example, values could be:
```
"region": "AZ",
"wifiName": "Bikini.Bottom.Wifi.2.4",
"wifiPassword": "5p0ng3B0b"
```

The call to `smartlink()` is made like this
```
constructedToken = region + token + secret // AZQyMLEFzw0NiE
TuyaSmartLink.smartLink(wifiName, wifiPassword, constructedToken, 5, 2, 1000, 1, 1);
```

**Comment by Ericmas001** [comment]
Oh and I do have some kind of conclusion knowing all of this: the url of the API is not sent to the device. Meaning, the device cannot register differently if it comes from SmartLife, Tyua, Jinvoo or eFamilyCloud. 

So the Apps are clearly all the same, since the device is clearly gonna call the same API with no distinction  depending witch App you are using. So my theory is when you register into one of those app, a user is created in the Global Database, but this user is associated with the app that you use. When you register a device, you are given a token/secret associated to your user, and by association to the App you are using. The device will be added to the Global Database, but linked to your user, so also by association linked to the App you are using. That's why user and device don't crossover between apps.

That means @codetheweb that we cannot send a custom server address to the device 😞 

The only thing I don't get is the need to have different APIs for the different Apps. They should call the same Tuya API with different clientId, and it should be enough.

**Comment by clach04** [comment]
@Ericmas001  dumb question, does spoofing the API address work? E.g. setup DNS server on your network to point to your server?

**Comment by Ericmas001** [comment]
I did not try. I guess the device is calling the API via https and this will fail because of the certificate, but I may be wrong

**Comment by BillSobel** [comment]
@Ericmas001 Did you happen to get any farther on the registration packets?  Looking at the header files for the wifi sdk I think you are seeing different registration steps as defined here:

typedef enum {
	SMARTLINK_STEP_HEAD = 0,
	SMARTLINK_STEP_MAGIC,
	SMARTLINK_STEP_PREFIX,
	SMARTLINK_STEP_DATA,
	SMARTLINK_STEP_FINISH,
	SMARTLINK_STEP_END,
	SMARTLINK_STEP_TLINK,
} TYPE_SMARTLINK_STEP;

**Comment by BillSobel** [comment]
Notes on AP registration steps:

Packet is UDP broadcast packet sent on 6669

int prefix 0x000055aa
int version 0x00000000
int command 0x00000001 (AP register?)
int size (json data token, passed, ssid)
json data (no null terminator)
crc32 (calculated from packet start to packetsize-8 [e.g. to end of json data])
int suffix 0x0000aa55

Device accepts packet and if wifi credentials are ok will connect to that new network.  I believe at that time it will validate the token, but I am off to confirm.  I wanted to simply leave some notes for others that might be following this down.

**Comment by BillSobel** [comment]
Some more notes, registration (for AP mode to start) is fairly straight forward:

Need cloud access keys from the developer portal (took me a few days to get approved)
Create a device token (tuya.m.device.token.create)
Turn on access point mode on device (LED flashes slowly)
Connect to device access point
Send UDP packet defined above
Connect to normal access point
If the normal access point has internet access, the device will connect and register itself with the cloud.  The calls to get a new UID and to activate a device are called by devices (apparently) and NOT the gui manager (so you can ignore these calls)

Call list token until the device you registered is returned as one of the items (or you hit a timeout you define)

That result will give you the deviceId.  You can call tuya.m.device.get with that deviceId to get the local encryption key.

Also once connected to the normal wifi the device will beacon every few seconds on UDP 6666.  This packet contains the deviceId (but not the local key).  

Thats it, device is now controllable.

**Comment by tjfontaine** [comment]
coincidentally I came to similar conclusions regarding the message layout, though not all of it entirely aligns with yours https://github.com/tjfontaine/tuyapi/commit/7f98bf05cff31fb704786a742a7d7f03ba21ffca

**Comment by BillSobel** [comment]
@tjfontaine I have to go review, I think what I marked as 'version' may actually be the packet id (other people saw this increment by 1 or a few numbers on each new packet).  Was there other differences? If so can you mention what you saw so we can compare?

**Comment by sebkajeka** [comment]
Would it be possible to use a custom app to register a device? This app could easily capture and display required information to use the API.

**Comment by codetheweb** [comment]
@sebkajeka yes, I'd like to.  The initial version would be a command line program, though, not an app for your phone.

**Comment by zuik** [comment]
@sebkajeka It seem to require the app key and app secret according to the [Tuya API](https://docs.tuya.com/en/app/app-sdk/android-sdk.html). However, I haven't been successful in reproducing their example

**Comment by sebkajeka** [comment]
Simplifying the device registration/localKey process is important. Am I correct in thinking it's possible to collect this information via API provided the device is registered with the Tuya App and we have our own app API credentials?

**Comment by codetheweb** [comment]
@sebkajeka no, devices are tied up to the keys they were registered with.  If you want to access the local key you need to register it yourself with your own API keys.

**Comment by BillSobel** [comment]
@sebkajeka As CodeTheWeb said, each developer set of keys is it's own walled garden in their backend.  So my@email.com with keyset1 can not see devices from the same my@email.com with keyset2.  So once you have a keyset all apps would have to use it (which would preclude using the smart life app)  When you register keys they will build you custom copies of those apps, it runs about $1600 all said and done (so I have not done that yet with them, not sure if I will bother as most people using devices via HomeSeer will be using ImperialHome or some other dashboard apps which should alleviate the need for their app)

@zuik I've got registration working, what part are you stuck at?

**Comment by zuik** [comment]
@BillSobel I tried following their example app at: https://github.com/TuyaInc/tuyasmart_android_sdk but it doesn't seem to compile properly in Android Studio. 
I don't know any Chinese, so navigating their website is a bit ...
Did you register as a developer to get the app key and app secret to put in your `TuyaSDK.init()` method?

**Comment by BillSobel** [comment]
@zuik Aww, sorry didn't realize you were using the SDK.  I wrote the code to do registration completely separate from the SDK (get the token, send the udp packet, wait for device to register itself).  Im in c# writing a HomeSeer plugin, so couldn't use any of the SDK.  The said, I did dig into (wrote some xposed hooks to look at some of their process).  I did register with them and get official keys from the, took about a week.  Also on the site, change the /cn/ in the URL to /en/ many of the pages are translated...

**Comment by zuik** [comment]
@BillSobel I see. I was reading through the decompiled java code and the biggest part that I was stuck on is how they do the "Fast Connect Mode" (The other one not the AP mode). In this thread, we know that they send out a bunch of UDP broadcast packets but most of these packets are zeros. 

I was wondering how they encode these into UDP packets? Did you use the AP mode or the "Fast Connect Mode"? How did you encode the WiFi credential into UDP packets?

**Comment by BillSobel** [comment]
@zuik I have not understood 'Fast Connect Mode' yet.  When a new device comes into an environment to be registered it needs the ssid and the password.  I have not been able to understand how the fast connect mode works given that.  Ive considered they may be using bluetooth, I know iOS has a special wifi setup mode that seems to open a second ssid concurrent with the primary one (assuming Android has something similar), that fast connect mode only works for a network the device had been on prior (so if you reset it, it still attempts to connect to the last network), that the device attempts to connect to the strongest open radio it sees, and lastly that the devices have an internal mesh so once one is up, the mesh can send the data to others.  So far I have not found anything conclusive on this so have only implemented the AP registration mode.  That mode is actually VERY simple to write (albeit it requires keys).

You do a cloud token request which gives you a token and a secret.  You setup a packet (just like a command packet, same structure) with the token field out as a string of the region+token+secret, the passwd field (wifi password) and the ssid field (ssid name).  You then broadcast that on port 6669 (not the same port as TCP).  The device does not respond.  If simply validates the packet (so the CRC needs to be correct) and extracts the data and connects to the AP.  If it connects and has open internet access it registers with the cloud.  You then poll the cloud on the token you gave the device and get back a list of devices which have registered with it (the token is good for 10 minutes so you can reuse it during that time, albeit probably easier to just get a new one for each registration you are doing, since this is AP mode you can only do one device at a time).

While we would all prefer this was 100% local, once that is in place you really don't need to talk to the cloud again, you have the local key and can control the device easily.

**Comment by sebkajeka** [comment]
Afaik, Fast Connect, EZ setup, SimpleConfig etc is just a variation of "SmartConfig" which is available on ESP8266. There's a few names for it but effectively what it does is encode the data in the length of a UDP packet. This way non-connected devices can listen to all nearby traffic for the appropriate change in packet lengths and start listening for configuration data. http://www.stijnvandrunen.nl/2015/06/esp8266-mqtt-smartconfig/

**Comment by Ericmas001** [comment]
That's also the conclusion I had that the packet length was the data
itself. I had decompiled the dll used by the app to try to understand how
it was encoded, but reading decompiled c++ code is... À different kind of
fun... :) your link is really helpful :)

**Comment by BillSobel** [comment]
@sebkajeka Thank you.  Packet length is something that never occurred to me.  I implemented a version of the sender from the project linked above.  Just need to determine the length encoding algorithm (hoping it is the same) and the data packet structure.  I tried a test using the same data as goes with an AP registration (json formatted) and no luck.  

If anyone has pointers on the data format or data to length algorithm please let me know, looks like Wireshark is going to be my friend for a few days!

**Comment by BillSobel** [comment]
@sebkajeka  @Ericmas001 I have figured out most of the data sent during the SmartLink process.  

The start is just a loop of packets of lengths 1, 3, 6, and 10.

As discussed the data is encoded into the length.  It appears there is an unknown 4 packet header (I am trying to figure out what these 4 lengths represent) followed by a sequence of 6 packets (this sequence repeats the same basic structure until the end of the data)

In the sequence of 6 packets, first packet is crc8 of the sequence number (next packet) and 4 data packets. The crc is done before the data is or'd to rebase it at 128 or 256.  The crc and the sequence number are modded by 128 and then based at 128 (finalcrc = (crc % 128) | 128.  

the data packets are simply the data or'd with 256 (so character 45 becomes 301).  Data is padded with 0 (which becomes 256) if the last group of packets is short (last group would be short 1-3 packets if the data length is not evenly divisible by 4).

The data itself is a length byte, the password bytes, length byte, token string (need to verify but suspect it is the region + token + secret string same as in other use cases), length byte, ssid name bytes.

So, hopefully just those header bytes to figure out, and this might actually work!

**Comment by codetheweb** [comment]
@BillSobel from what you've seen so far, do you think it would be possible to generate a random local key and just send that to the device?  Or does it need a key registered with Tuya's API to work properly?

**Comment by BillSobel** [comment]
@codetheweb Definitely requires the cloud to configure the key.  You send the device a token the cloud generates which you then both use to 'find' each other, no options to send it a key or other material to use.  Once configured you can run fully locally, but until then no....

**Comment by BillSobel** [comment]
An update.  The header is just the length of all the strings (+2) encoded 4 different ways.  The CRC8 calculation doesn't match a normal CRC8 (or any of the known ones I can find documented).  But the code from the SmartLink android project mentioned above does implement the same CRC8 and works for this.  

//
            // Fill in first 4 bytes 
            // of header here
            //
            var stringLength = (wifiNameBytes.Length + regionTokenSecretBytes.Length + wifiNameBytes.Length + 2) % 256; ;
            var stringLengthBuffer = new byte[1];
            stringLengthBuffer[0] = (byte)(stringLength);
            var stringLengthCrc = CalculateTuyaCrc8(stringLengthBuffer, 1);

            // Length encoded into the first two bytes based at 16 and then 32
            encodedData[0] = (stringLength / 16) | 16;
            encodedData[1] = (stringLength % 16) | 32;
            // Length CRC encoded into the next two bytes based at 46 and 64
            encodedData[2] = (stringLengthCrc / 16) | 48;
            encodedData[3] = (stringLengthCrc % 16) | 64;

**Comment by BillSobel** [comment]
It works!  I got an outlet device to register to the cloud by doing the SmartLink protocol to send the config data over.  So now we know we can do both Smartlink and AP registration modes (albeit both need the cloud involvement)

**Comment by sebkajeka** [comment]
Well done! So to confirm, we need a Cloud API key from developer.tuya. With the cloud API we can now register devices and control them via the same API?

Are you publishing this code somewhere @BillSobel?

**Comment by BillSobel** [comment]
@sebkajeka The actual code is part of a closed source project, so I cant post it all, but can definitely share the code around registration.  I wouldn't have gotten here without this group, and more than happy to ensure others can do the same thing.

**Comment by codetheweb** [comment]
@BillSobel could you please update this list with corrections?  I had a hard time understanding some of what you said.

**Beginning Pattern** (loops for 50+ iterations)
1. Packet of length 1
2. Packet of length 3
3. Packet of length 6
4. Packet of length 10

**Data Pattern** (loops until all data has been transmitted)
- Header
  - **Calculated by**
  ```
  var stringLength = (wifiNameBytes.Length + regionTokenSecretBytes.Length + wifiNameBytes.Length + 2) % 256; ;
  var stringLengthBuffer = new byte[1];
  stringLengthBuffer[0] = (byte)(stringLength);
  var stringLengthCrc = CalculateTuyaCrc8(stringLengthBuffer, 1);
  // Length encoded into the first two bytes based at 16 and then 32
  header[0] = (stringLength / 16) | 16;
  header[1] = (stringLength % 16) | 32;
  // Length CRC encoded into the next two bytes based at 46 and 64
  header[2] = (stringLengthCrc / 16) | 48;
  header[3] = (stringLengthCrc % 16) | 64;
  ```
  - **Transmitted as**
    1. `header[0]`
    2. `header[1]`
    3. `header[2]`
    4. `header[3]`
- Data
  - **Calculated by**
    - Length (of password) byte
    - Password bytes
    - Length (of token string) byte (what is the token string?)
    - Token string
    - Length (of SSID) byte
    - SSID bytes
  - **Transmitted as**
    1. CRC8 of next packet's sequence number? (UDP packets don't have sequence numbers)
    2. Placeholder packet for above?
    3. `dataByte[0] | 256`
    4. `dataByte[1] | 256`
    5. `dataByte[2] | 256`
    6. `dataByte[3] | 256`

**Comment by codetheweb** [comment]
Ok, thanks for the code @BillSobel.

**Comment by BillSobel** [comment]
@Ericmas001 Sort of, you can think of it logically as one big data base per region but the actual implementation is probably different.  Region is the first main key, that actually drives you to one of three amazon data centers (europe, us, or asia).  Within that data center logically all devices are registered but the same device registered with different keys will basically be treated as completely different devices.  

When you get a token, the token is valid within your namespace on their backend, so when the device registers with it, they know whom to assign it to.

**Comment by codetheweb** [comment]
Hey @BillSobel, would you mind looking at my code [here](https://github.com/codetheweb/TuyaRegister)?  I translated your protocol implementation from C# to Javascript, but it's not working for some reason, even though captured packets from both my program and the app look identical.

The only thing I can think of is that the timing between packets and packet groups might be off (is that important?).

**Comment by BillSobel** [comment]
@codetheweb The code referenced earlier which got me started on this had those options to change up the packet delay, but you copied them so you should be good.  Have you verified data is on the wire in the expected format?  The code looks correct, but I have one (possibly naive) question on when you calculate the CRC:

let stringLengthCRC = this.tuyaCRC8([stringLength], 1);

I'm assuming that stringLength is an int.  Are you sure the crc function is getting the right byte of the int (the low byte of the low word) in this calculation?  It appears to be either indeed getting it right or potentially getting the high byte of the high word.  The second thing is, while the CRC implimentation looks right, do you want to crc a few example strings and do the same (or I can if your not setup) in the c# version to ensure its indeed calculating the same results.

**Comment by BillSobel** [comment]
How about the crc of the int, did you verify the right byte is getting the crc calc applied?

**Comment by codetheweb** [comment]
I'm not really sure what you mean by the 'right byte' @BillSobel.  I could be wrong, but from my (limited) knowledge of high/low bytes and the way JS treats variables, I believe `stringLength` is only stored as one byte.  I could be wrong.  However, `Buffer.from([stringLength]).readInt8(0) === stringLength`.

**Comment by BillSobel** [comment]
@codetheweb That sounds correct, so Im unfortunately not seeing any obvious errors in the port.  Any chance you can wireshark the conversation?

**Comment by codetheweb** [comment]
[Here's](https://file.io/wuNemf) a capture with these parameters @BillSobel:
- Region: `AZ`
- Token: `asdfghjk`
- Secret: `asdf`
- SSID: `asdf`
- WiFi password: `asdf`

Thanks for all your help.

**Comment by BillSobel** [comment]
@codetheweb Just get a 404 on the link...

**Comment by codetheweb** [comment]
Sorry about that, try [this](https://www.cloudshark.org/captures/1f2c132b23ed) @BillSobel.

**Comment by BillSobel** [comment]
The first 4 bytes of your header (from the capture) are: 17,40,79.53   While I don't see an issue yet in the code, 79 doesn't appear valid.  Since the CRC is a single byte, it its range is 0-255.  /16 give us a range 0-15.  Its then based at 48 giving a range of 48-63.  I don't see how the 79 is getting in there unless the CRC function is returning a value > 256.

**Comment by BillSobel** [comment]
Also the rest of the data is misordered.  The packet sequence should appear every 6 bytes (CRC, seq, 4 data bytes).  The packet capture shows the seq # 128, 129, 130, etc but they are jumbled through the set for some reason.

**Comment by codetheweb** [comment]
I get what you're saying about the invalid header byte, but a packet capture of the app also yields packets of length 79.
The packet order does appear to be jumbled, I'll take a look at that.

**Comment by BillSobel** [comment]
The next byte is 53, which should be a minimum of 64.  I expect those two bytes are just swapped as we are seeing elsewhere, so looking like one issue not two.

**Comment by codetheweb** [comment]
The root issue appears to be that the bytes are correctly ordered in memory, but are sent out in the wrong order for some reason.  Still trying to figure out why this is happening.

**Comment by Ericmas001** [comment]
Took a look at your code and i'm in no way experimented with javascript and node.js, so i'm probably wrong, but in the following code:
```data.forEach(async function(b) {
        await that.broadcastUDP(b);
        await delay(delayMs);
      });```

Is it possible that the foreach will fire the function, then iterate to the next one, without waiting for the function to end ?

**Comment by BillSobel** [comment]
@Ericmas001  Im pretty sure its a variation of that.

data.forEach(async function(b) {
        await that.broadcastUDP(b);
        await delay(delayMs);
      });

It appears that each byte gets launched in a lamda, each lamda correctly awaits and delays, but I think the lamda itself needs to be awaited otherwise you are basically launching them all in parallel.  I don't know the proper syntax here but this article discussing the issue and provides solutions:

https://codeburst.io/javascript-async-await-with-foreach-b6ba62bbf404

**Comment by BillSobel** [comment]
@codetheweb (from article) try:

async function asyncForEach(array, callback) {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array)
  }
}

**Comment by codetheweb** [comment]
You're right, it was that function.  The packets are now in the correct order, but it's still not working.  [Here's](https://www.cloudshark.org/captures/07020df2485d) a capture with the same parameters as before.
If everything looks right in the capture, it could be my device at fault and not my program as I've begun to have trouble when trying to add it to the official app.

**Comment by BillSobel** [comment]
@codetheweb The latest trace looks correct.  Question, are you testing with that actual secret string, or are you getting one from the cloud and using that?  Im not sure if the device does any internal validation on the region, token,secret other to make sure its 14 characters, but if it is and your not using real data it could be ignoring it.

**Comment by codetheweb** [comment]
I'm using real data.  [Here's](https://gist.github.com/codetheweb/93d0876ec833aa31ce7c4f0c90034b69) a gist of the code I'm using to generate the token and secret (I borrowed the bulk of the code from [bahorn](https://gist.github.com/bahorn/160b4143badd1b6fae61cec629fce339)).

**Comment by codetheweb** [comment]
@BillSobel If you're able to, can you try generating a packet capture with these parameters:
- Region: AZ
- Token: asdfghjk
- Secret: asdf
- SSID: asdf
- WiFi password: asdf

If your capture is different than mine, then we know my code's at fault.  Otherwise, it must be the device.

**Comment by sebkajeka** [comment]
Did it match?

**Comment by codetheweb** [comment]
For the most part.  2 or 3 packets didn't match, but I haven't had the time to look into it yet.

However, my device didn't respond when I used the C# version, so at this point I think there's something wrong with my plug.  I've ordered a new one.

**Comment by Ericmas001** [comment]
Does the plug works with the official app?

**Comment by codetheweb** [comment]
Ok; I just received my new plug today and it's still not working.

If you don't mind me asking, how many characters were in the SSID and password that it worked with @BillSobel?

**Comment by BillSobel** [comment]
Definite bug. Something off with last packet.  Out but next on list when back.  Give me 12-18 hours and I’ll hsve an update.

**Comment by BillSobel** [comment]
@codetheweb Please change the rounder function per the below.  Murphy's law and all, I tested against 3 SSID/PW combos and none hit the issue (last packet wasn't rounding up properly).  The total data length was (as an example in my new case) 45, should have round to 48 but was returning 44.  So last character dropped.  Made this change and the builds I was testing linked straight off.

        private static double Rounder(double x, double g = 4)
        {
            return Math.Ceiling(x / g) * g;
        }

**Comment by codetheweb** [comment]
I literally just figured out the same thing when I received a notification email of your new comment :).

It seems to now correctly connect to the network, but how is the `localKey` derived from the token (and secret)?

**Comment by codetheweb** [comment]
Yeah, I'm trying but it doesn't appear my device is correctly registering with the cloud...

**Comment by BillSobel** [comment]
Does the ssid have internet access? Device should stop flashing quickly as it attempts to enroll.

**Comment by codetheweb** [comment]
Ok, here's what's currently happening:
 - After sending valid token and secret to device, device:
    - Stops flashing
    - Connects to network at a pingable address
    - Contacts Tuya servers (seen by inspecting DNS logs for my network)
    - Turns solid blue
   - *However, calling `tuya.m.device.list` returns nothing.*

- After sending invalid (random) token and secret to device, device:
  - Stops flashing
  - Connects to network
  - Contacts Tuya servers
  - Goes back to flashing once it realizes the token is incorrect

Any ideas as to what's going on?

**Comment by BillSobel** [comment]
Are you using the same key to get the token as to call get devices? Did you call token for device?

**Comment by codetheweb** [comment]
Yes, I'm using the same API key and secret for all requests.

I've been trying `tuya.m.device.list`, but I just tried polling `tuya.m.device.list.token` (which is I think what you mean) with the generated token as an argument and still nothing.

I even tried registering it over a hotspot to migrate any potential network problems, and still no luck.  This is getting frustrating, because the device is clearly connecting to a network but I can't control it without the localKey.

**Comment by BillSobel** [comment]
Ate the keys real? Eg assigned by tuya? Also is the zone used for logging in the same zone you use when getting and constructing the token? US severs can’t see EU registrations so the zone needs to match.

**Comment by codetheweb** [comment]
Finally got it working!  Apparently I had a stray space character in the `timeZone` parameter (used when creating a new token).  Weird that Tuya would return without an error, but whatever.  Thanks for all your help and support @BillSobel.

@ everyone I'll be cleaning up my code, converting the cloud API module from Python to JS, and making a CLI tool for registering devices over the next few weeks.  Follow my progress (and help out) [here](https://github.com/codetheweb/TuyaRegister).

**Comment by BillSobel** [comment]
@codetheweb The token probably wound up in a different zone, hence the completion of registration but the missing device when looking for it.  Glad you up and running! That was fun to work on :)  

I noticed some things in the API around groups of devices and meshing, I am assuming these are logical groups but at the same time curious to see if the devices read from a different MQTT queue under a group ID or if they simply sync multiple lights by assuming each will read 'fast enough' so that users won't see them changing independently.  If you see anything in your splunking, please let me know.

**Comment by codetheweb** [comment]
**Update**: I've finished translating, refactoring, testing, and documenting a module for calling Tuya's cloud API.  It's published under a new scoped organization, `tuyapi`.  You can find the package [here](https://www.npmjs.com/package/@tuyapi/cloud) and the source [here](https://github.com/TuyaAPI/cloud).  The source is also under a new Github Organization.

Going forward, I plan to create at least two more packages (both under the `tuyapi` Github/NPM organization):
- `register` - this includes the low level UDP code for registering devices as well as functions that rely on `@tuyapi/cloud` for generating tokens
- `cli` - this package simply wraps `@tuyapi/register` and `codetheweb/tuyapi` in a CLI interface for bash scripts and to make registering devices as simple as possible.  Probably will end up being invoked by `tuya-cli register ...` or similar.

I would like to move `codetheweb/tuyapi` to `@tuyapi/base`, but I'm still trying to figure out the possible issues that would arise from that change.

Any feedback on the above stuff is great.

@sebkajeka I'll be sure to document the process somewhere in the `register` package. :)

**Comment by jimburkegb** [comment]
Reading these very good, informed posts suggests that this is the place to post my problem.

I have a network problem when I try to install two iGenix panel heaters on my home network. Other smart devices have connected and function correctly, albeit with some initial problems.

In short, the configuration sequence between device and app running on mobile phone does not complete. It starts fine, but times out with an error.
In detail, the sequence is :-
1. Installed app on mobile phone
2. Switched iGenix heater to config mode (WiFi LED flashing quickly)
3. Started app
4. On app, switched to SSID of nearby Draytek hub
5. On app, entered WiFi password 
6. On app, started configuration. (WiFi LED still flashing quickly)
7. Spinner gets to 10% and panel heater's WiFi LED goes out.
8. Spinner gets to 20% and panel heater's WiFi LED comes on, no flashing.
9. At this point, the heater appears in the DHCP table and pings fine.
10. Spinner goes all the way round to 100% and shows error.

I've repeated the sequence on both heaters. 
I've used TWO apps - iGenix and SmartLife - both have the same behaviour.
I've used another phone - same problem.
There's plenty bandwidth - I switched off three security cameras.
I switched the WiFi hub to 2.4 GHz and disabled 5 GHz.
WPA/WPA2-PSK enabled.
Tried AP mode - same problem. I note that the 'private' SSID of the heater appears - "SmartLife_D7F1".
In the DHCP table, the devices appear as "EPS_DED7F1" and "EPS_DB69B7". I see that the last six digits are the same as the last six digits in the MAC address. They have IP addresses assigned and they ping fine.

Now then, I tried to break down the problem by getting another mobile phone and switched it to a hotspot to which I connect my mobile phone. [This effectively excludes my WiFi hub and (Virgin) router] 
I went through the configuration sequence and it worked !!!  This indicates that the phone works fine, the heater work fine.

Can anyone help ?

Many thanks,
Jim Burke
